### PR TITLE
3D printer board Big Tree Tech EBB42, missing Generic STM32G0 and STM32G4 support

### PR DESCRIPTION
**Pull request by alextrical**
Added new Generic support for STM32G0B1/C1 support, and 3D printer board Big Tree Tech EBB42 CANBUS V1.1 motor / hot end control board

### CODE REVIEW

File: variants/STM32G0xx/G0B1C(B-C-E)(T-U)_G0C1C(C-E)(T-U)/generic_clock.c
Comment (ABOSTM) [inline variants/STM32G0xx/G0B1C(B-C-E)(T-U)_G0C1C(C-E)(T-U)/generic_clock.c:28]: In order to remove compilation warning, please remove '0' like below
```suggestion
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]
1 small comment, otherwise LGTM

**Review by ABOSTM — APPROVED** [review]
LGTM, Thanks

File: variants/STM32G0xx/G030J6M/generic_clock.c
Comment (ABOSTM) [inline variants/STM32G0xx/G030J6M/generic_clock.c]: Please remove `0` to avoid compilation warning.
The same for all clock configuration.

**Review by ABOSTM — CHANGES_REQUESTED** [review]
@alextrical,
Thanks, that is a huge work.
Few comments:
* Missing Readme update corresponding to the new MCUs.
* Could you run Astyle to harmonize formatting:
https://github.com/stm32duino/wiki/wiki/Astyle#python-script

File: variants/STM32G0xx/G030J6M/generic_clock.c
Comment (ABOSTM) [reply variants/STM32G0xx/G030J6M/generic_clock.c]: This needs to be done on both

Comment (alextrical) [reply variants/STM32G0xx/G030J6M/generic_clock.c]: No worries, i will try to do that now. sorry if multiple Pull requests have been made, I though there was another button i needed to select to submit (and went looking multiple times for it)

Comment (alextrical) [reply variants/STM32G0xx/G030J6M/generic_clock.c]: Not sure why the 0's and Readme issues are still there, i checked those fixes in, in the last hour

**Comment by alextrical** [comment]
> * Could you run Astyle to harmonize formatting

I'm going to try that now, Once I download the ".astyleignore", ".astylerc" and "astyle.py" files, Do i run the py file at the root of the project, and do i need to manually specify in the command to use the definition files?

File: variants/STM32G0xx/G030J6M/generic_clock.c
Comment (alextrical) [reply variants/STM32G0xx/G030J6M/generic_clock.c]: apologies for the issues i'm causing, i'm still trying to get used to GitHub when working with Pull Requests. Hopefully i will get the hang of it soon

**Comment by ABOSTM** [comment]
In fact scripts to ease Astyle are already available when cloning github:
https://github.com/stm32duino/Arduino_Core_STM32/tree/main/CI/astyle

But you need to download Astyle itself.


Then all we request is to run Astyle on the sources/headers files you modified.
So you can run for example :
`> python3 CI/astyle/astyle.py  -p  /c/Arduino/AStyle/bin -r ./variants/STM32G0xx/`
(suppose Astyle is installed in /c/Arduino/AStyle/bin)

**Comment by alextrical** [comment]
I've run Astyle with the following command 
`python astyle.py -i .astyleignore -d .astylerc -r /var/home/alextrical/GitHub/Arduino_Core_STM32/variants/STM32G0xx
`
It looks to have made the required changes to the 'generic_clock.c' files in each file that has been created for this pull :)

Hopefully the requested changes are all now showing up.
i.e. Astyle formatting of 'generic_clock.c' / files
Values removed from the following for all boards added : RCC_ClkInitStruct = {}; & RCC_OscInitStruct = {};
Readme updated for all MCU's

I will remember these for the next batch ;) and leave it a bit longer between making the changes, and placing a change request to hopefully let it capture the full set of changes, and hopefully i will only create one pull request next time ;)

**Review by SornakumarN — APPROVED** [review]

File: variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c
Comment (ABOSTM) [inline variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c:27]: Some remaining '0'

File: variants/STM32G0xx/G031G(4-6-8)U_G041G(6-8)U/generic_clock.c
Comment (ABOSTM) [inline variants/STM32G0xx/G031G(4-6-8)U_G041G(6-8)U/generic_clock.c:25]: dito

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Some remaining '0' to be removed, including in variant_EBB42_v1_1.cpp

File: variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c
Comment (alextrical) [reply variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c:27]: How is notepad ++ missing those. Will find a RegEx to scan for them
Thank you for spotting these, sorry for missing them

File: README.md
Comment (alextrical) [reply README.md:318]: I've only added the ":yellow_heart:" lines, I don't believe that I have added anything to the variants that's already in the list / Green_heart. 
Its possible that the process of putting the list around there in alphabetical order has triggered the change tracking to show an addition that I didn't make to the underlaying variants folders

File: variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c
Comment (alextrical) [reply variants/STM32G0xx/G031C(4-6-8)(T-U)_G041C(6-8)(T-U)/generic_clock.c]: Fixed

File: variants/STM32G0xx/G031G(4-6-8)U_G041G(6-8)U/generic_clock.c
Comment (alextrical) [reply variants/STM32G0xx/G031G(4-6-8)U_G041G(6-8)U/generic_clock.c]: resolved

File: README.md
Comment (alextrical) [reply README.md:312]: Not a new board, it was in the wrong alphabetical order, and moved from line 308. J shouldn't be before F

File: README.md
Comment (alextrical) [reply README.md:318]: Not a new board, it was in the wrong alphabetical order, and moved from line 312. J shouldn't be before F

**Comment by alextrical** [comment]
How do I go about creating a separate commit, I was hoping the last Pull request for the G4 would have been a separate entity from the G0 submission. Is the Close with comment enough?

**Comment by alextrical** [comment]
once this pull request is closed I will start adding a few other definitions, that are queued up and ready to go. Currently sitting in another repo for safe storage ready to be added.
https://github.com/alextrical/STM32-linker-scripts/tree/main/variants

**Comment by ABOSTM** [comment]
Sorry, I mean separate Pull Request :smile: 
Yes, PR needs extra check, and at the end PR needs to be `merged` to be fully checked in.
Extra check includes CI automatic check which are currently failed (see above)

At first glance there is an error while compiling GENERIC_G0B0CETX, GENERIC_G0B0RETX  and GENERIC_G0B0VETX  
/github/home/.arduino15/packages/STMicroelectronics/hardware/stm32/2.3.0/cores/arduino/stm32/uart.h:120:21: error: ```
'USART3_4_IRQn' undeclared (first use in this function); did you mean 'USART4_IRQn'?
  120 | #define USART3_IRQn USART3_4_IRQn
```
I will have a look at this issue

**Comment by alextrical** [comment]
ah, well that's a pain that its not compiling :/
I can see why this should be submitted as multiple smaller Pull requests 😅
unfortunately i made the changes directly into the Main branch of my fork, and now I don't know if its possible to create a branch without causing extra issues for future PR's

**Comment by ABOSTM** [comment]
I made a fix for the compilation error #1790.
So I prefer you don't remove the failed MCU and thus have a complete G0 support.
Whatever, we will wait for @fpistm to review before merging.

Nevertheless, you can push other STM32 series in parallel (one per PR would be great).
Warning each PR should be independent one from each other, as long as they are not merged.
Exemple: if you push a new PR to add MCU for F1, it should not be on top of current G0 PR (as long as it is not merged) nor other potential PR for F0 series you have in the pipe ...

Alternatively, you can wait for this one to be merged, before pushing next PR. This will take more time to reach the end of STM32 series, but it is not an issue.
Please note that next week, support will be drastically reduced due to summer holidays.

**Comment by alextrical** [comment]
That's fair enough, I will leave the G0B0xx series in place.

As for pushing independent PR's, I'm guessing that needs to be done with branches within my copy of the project, with 1 branch per MCU family. Unfortunately I did all previous modifications directly in 'Main' so have somewhat contaminated the base of my fork

**Comment by ABOSTM** [comment]
> As for pushing independent PR's, I'm guessing that needs to be done with branches within my copy of the project, with 1 branch per MCU family.

Right

> Unfortunately I did all previous modifications directly in 'Main' so think I have somewhat contaminated the base of my fork

Thus let's wait for merge of 1st PR before pushing next one.

**Review by ABOSTM — CHANGES_REQUESTED** [review]
*2.4.0*

**Review by alextrical — COMMENTED** [review]
Version Number now updated to show 2.4.0 in readme.md

**Comment by alextrical** [comment]
Hopefully that is now in sync with the main repo. Please give the CI a try

**Comment by fpistm** [comment]
Hi @alextrical 
I've clean up the branch as I could not simply squash it. Now only 3 commits.

**Review by fpistm — APPROVED** [review]
LGTM

**Pull request merged**
